File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/models/sed/nav/Person.java

package no.nav.melosys.eessi.models.sed.nav;

import java.util.Collection;
import java.util.HashSet;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import lombok.Data;


@JsonInclude(Include.NON_NULL)
@Data
public class Person {

    private String etternavn;

    private String etternavnvedfoedsel;

    private Foedested foedested;

    private String foedselsdato;

    private String fornavn;

    private String fornavnvedfoedsel;

    private String kjoenn;

    private Collection<Pin> pin = new HashSet<>();

    private Collection<Statsborgerskap> statsborgerskap = new HashSet<>();

    public Optional<Pin> finnNorskPin() {
        return pin.stream().filter(p -> "NO".equals(p.getLand())).findFirst();
    }

    public Collection<String> hentStatsborgerksapsliste() {
        return statsborgerskap.stream()
                .filter(Objects::nonNull) //Kan av en mystisk grunn inneholde null-felter
                .map(Statsborgerskap::getLand)
                .collect(Collectors.toSet());
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/models/sed/nav/Person.kt
package no.nav.melosys.eessi.models.sed.nav

import com.fasterxml.jackson.annotation.JsonInclude
import no.nav.melosys.eessi.identifisering.FnrUtils
import java.util.*
import kotlin.jvm.optionals.getOrNull

@JsonInclude(JsonInclude.Include.NON_NULL)
data class Person(
    var etternavn: String? = null,
    var etternavnvedfoedsel: String? = null,
    var foedested: Foedested? = null,
    var foedselsdato: String? = null,
    var fornavn: String? = null,
    var fornavnvedfoedsel: String? = null,
    var kjoenn: Kjønn? = null,
    var pin: List<Pin> = emptyList(),
    var statsborgerskap: List<Statsborgerskap?> = emptyList()
) {
    fun hentStatsborgerksapsliste(): Collection<String> = statsborgerskap.mapNotNull { it?.land }

    fun harStatsborgerskap(land: String): Boolean = hentStatsborgerksapsliste().contains(land)

    fun finnUtenlandskIdFraLand(land: String): Optional<Pin> = pin.firstOrNull { it.land == land }?.let { Optional.of(it) } ?: Optional.empty()

    fun hentNorskPersonnummer(): String? =
        pin.firstOrNull { it.land == "NO" }?.identifikator?.let { FnrUtils.filtrerUtGyldigNorskIdent(it).getOrNull() }

    fun harNorskPersonnummer(): Boolean = hentNorskPersonnummer() != null
}
